1.0.2/xo/xo.1
.\" #
.\" # Copyright (c) 2014, Juniper Networks, Inc.
.\" # All rights reserved.
.\" # This SOFTWARE is licensed under the LICENSE provided in the
.\" # ../Copyright file. By downloading, installing, copying, or 
.\" # using the SOFTWARE, you agree to be bound by the terms of that
.\" # LICENSE.
.\" # Phil Shafer, July 2014
.\" 
.Dd December 4, 2014
.Dt XO 1
.Os
.Sh NAME
.Nm xo
.Nd emit formatted output based on format string and arguments
.Sh SYNOPSIS
.Nm
.Op Fl options
.Op Ar argument...
.Sh DESCRIPTION
The
.Nm
utility allows command line access to the functionality of
the
.Nm libxo
library.
Using
.Nm ,
shell scripts can emit
.Em XML ,
.Em JSON ,
or
.Em HTML
using the same commands that emit text output.
.Pp
.Bl -tag -width indent
.It Ic --close Ar path
Close tags for the given path
.It Ic -C | Ic --continuation
Indicates this output is a continuation of the previous output data
and should appear on the same line.
This is allows HTML output to be constructed correctly.
.It Ic --depth Ar num
Set the depth for pretty printing
.It Ic --help
Display help text
.It Ic -H | Ic --html
Generate HTML output
.It Ic -J | Ic --json
Generate JSON output
.It Ic --leading-xpath Ar path
Add a prefix to generated XPaths (HTML)
.It Ic --not-first
Indicate that this content is not the first in a series of sibling
objects, which is vital information for "JSON" output, which requires
a comma between such objects.
.It Ic --open Ar path
Open tags for the given path
.It Ic -p | Ic --pretty
Make 'pretty' output (add indent, newlines)
.It Ic --style Ar style
Generate given style (xml, json, text, html)
.It Ic -T | Ic --text
Generate text output (the default style)
.It Ic --top-warp
Indicates the entire object should be placed inside a top-level
object wrapper, specifically when generating JSON output.
.It Ic --version
Display version information
.It Ic -W | Ic --warn
Display warnings in text on stderr
.It Ic --warn-xml
Display warnings in xml on stdout
.It Ic --wrap Ar path
Wrap output in a set of containers
.It Ic -X | Ic --xml
Generate XML output
.It Ic --xpath
Add XPath data to HTML output
.El
.Pp
The
.Nm
utility accepts a format string suitable for
.Xr xo_emit 3
and a set of zero or more arguments used to supply data for that string.
.Pp
In addition,
.Nm
accepts any of the
.Nm libxo
options listed in
.Xr xo_options 7 .
.Sh EXAMPLES
In this example,
.Nm
is used to emit the same data encoded in text and then in XML by
adding the "-p" (pretty) and "-X" (XML output) flags:
.Bd -literal -offset indent
  % xo 'The {:product} is {:status}\\n' stereo "in route"
  The stereo is in route
  % xo -p -X 'The {:product} is {:status}\\n' stereo "in route"
  <product>stereo</product>
  <status>in route</status>
.Ed
.Pp
In this example, the output from a
.Nm
command is shown in several styles:
.Bd -literal -offset indent
  xo "The {k:name} weighs {:weight/%d} pounds.\\n" fish 6
.Pp
  TEXT:
    The fish weighs 6 pounds.
  XML:
    <name>fish</name>
    <weight>6</weight>
  JSON:
    "name": "fish",
    "weight": 6
  HTML:
    <div class="line">
      <div class="text">The </div>
      <div class="data" data-tag="name">fish</div>
      <div class="text"> weighs </div>
      <div class="data" data-tag="weight">6</div>
      <div class="text"> pounds.</div>
    </div>
.Ed
.Pp
The
.Fl "-wrap <path>"
option can be used to wrap emitted content in a
specific hierarchy.
The path is a set of hierarchical names separated
by the '/' character.
.Bd -literal -offset indent
  xo --wrap top/a/b/c '{:tag}' value
.Pp
  XML:
    <top>
      <a>
        <b>
          <c>
            <tag>value</tag>
          </c>
        </b>
      </a>
    </top>
  JSON:
    "top": {
      "a": {
        "b": {
          "c": {
            "tag": "value"
          }
        }
      }
    }
.Ed
.Pp
The
.Fl "\-open <path>"
and
.Fl "\-close <path>"
can be used to emit
hierarchical information without the matching close and open
tag.
This allows a shell script to emit open tags, data, and
then close tags.
The
.Fl \-depth
option may be used to set the
depth for indentation.
The
.Fl "\-leading-xpath"
may be used to
prepend data to the XPath values used for HTML output style.
.Bd -literal -offset indent
  #!/bin/sh
  xo --open top/data
  xo --depth 2 '{tag}' value
  xo --close top/data
.Pp
  XML:
    <top>
      <data>
        <tag>value</tag>
      </data>
    </top>
  JSON:
    "top": {
      "data": {
        "tag": "value"
      }
    }
.Ed
.Sh SEE ALSO
.Xr libxo 3 ,
.Xr xo_emit 3 ,
.Xr xo_options 7
.Sh HISTORY
The
.Nm libxo
library first appeared in
.Fx 11.0 .
.Sh AUTHORS
.Nm libxo
was written by
.An Phil Shafer Aq Mt phil@freebsd.org .



NEWS
--
NTP 4.2.8p12 (Harlan Stenn <stenn@ntp.org>, 2018/14/09)

NOTE: this NEWS file will be undergoing more revisions.

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes a "hole" in the noepeer capability introduced to ntpd
in ntp-4.2.8p11, and a buffer overflow in the openhost() function used by
ntpq and ntpdc.  It also provides 26 other bugfixes, and 4 other improvements:

* [Sec 3505] Buffer overflow in the openhost() call of ntpq and ntpdc.

* [Sec 3012] Fix a hole in the new "noepeer" processing.

* Bug Fixes:
 [Bug 3521] Fix a logic bug in the INVALIDNAK checks.  <stenn@ntp.org>
 [Bug 3509] Add support for running as non-root on FreeBSD, Darwin,
            other TrustedBSD platforms
 - applied patch by Ian Lepore <perlinger@ntp.org>
 [Bug 3506] Service Control Manager interacts poorly with NTPD <perlinger@ntp.org>
 - changed interaction with SCM to signal pending startup
 [Bug 3486] Buffer overflow in ntpq/ntpq.c:tstflags() <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3485] Undefined sockaddr used in error messages in ntp_config.c <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3484] ntpq response from ntpd is incorrect when REFID is null <perlinger@ntp.org>
 - rework of ntpq 'nextvar()' key/value parsing
 [Bug 3482] Fixes for compilation warnings (ntp_io.c & ntpq-subs.c) <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3480] Refclock sample filter not cleared on clock STEP <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3479] ctl_putrefid() allows unsafe characters through to ntpq <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3476]ctl_putstr() sends empty unquoted string [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods); not sure if that's bug or feature, though
 [Bug 3475] modify prettydate() to suppress output of zero time <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3474] Missing pmode in mode7 peer info response <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3471] Check for openssl/[ch]mac.h.  HStenn.
 - add #define ENABLE_CMAC support in configure.  HStenn.
 [Bug 3470] ntpd4.2.8p11 fails to compile without OpenSSL <perlinger@ntp.org>
 [Bug 3469] Incomplete string compare [...] in is_refclk_addr <perlinger@ntp.org>
 - patch by Stephen Friedl
 [Bug 3467] Potential memory fault in ntpq [...] <perlinger@ntp.org>
 - fixed IO redirection and CTRL-C handling in ntq and ntpdc
 [Bug 3465] Default TTL values cannot be used <perlinger@ntp.org>
 [Bug 3461] refclock_shm.c: clear error status on clock recovery <perlinger@ntp.org>
 - initial patch by Hal Murray; also fixed refclock_report() trouble
 [Bug 3460] Fix typo in ntpq.texi, reported by Kenyon Ralph.  <stenn@ntp.org>
 [Bug 3456] Use uintptr_t rather than size_t to store an integer in a pointer
 - According to Brooks Davis, there was only one location <perlinger@ntp.org>
 [Bug 3449] ntpq - display "loop" instead of refid [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3445] Symmetric peer won't sync on startup <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3442] Fixes for ntpdate as suggested by Gerry Garvey,
 with modifications
 New macro REFID_ISTEXT() which is also used in ntpd/ntp_control.c.
 [Bug 3434] ntpd clears STA_UNSYNC on start <perlinger@ntp.org>
 - applied patch by Miroslav Lichvar
 [Bug 3426] ntpdate.html -t default is 2 seconds.  Leonid Evdokimov.
 [Bug 3121] Drop root privileges for the forked DNS worker <perlinger@ntp.org>
 - integrated patch by  Reinhard Max
 [Bug 2821] minor build issues <perlinger@ntp.org>
 - applied patches by Christos Zoulas, including real bug fixes
 html/authopt.html: cleanup, from <stenn@ntp.org>
 ntpd/ntpd.c: DROPROOT cleanup.  <stenn@ntp.org>
 Symmetric key range is 1-65535.  Update docs.   <stenn@ntp.org>

--
NTP 4.2.8p11 (Harlan Stenn <stenn@ntp.org>, 2018/02/27)

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes 2 low-/medium-, 1 informational/medum-, and 2 low-severity
vulnerabilities in ntpd, one medium-severity vulernability in ntpq, and
provides 65 other non-security fixes and improvements:

* NTP Bug 3454: Unauthenticated packet can reset authenticated interleaved
	association (LOW/MED)
   Date Resolved: Stable (4.2.8p11) 27 Feb 2018
   References: Sec 3454 / CVE-2018-7185 / VU#961909
   Affects: ntp-4.2.6, up to but not including ntp-4.2.8p11.
   CVSS2: MED 4.3 (AV:N/AC:M/Au:N/C:N/I:N/A:P) This could score between
	2.9 and 6.8.
   CVSS3: LOW 3.1 CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:N/I:N/A:L This could
	score between 2.6 and 3.1
   Summary:
	The NTP Protocol allows for both non-authenticated and
	authenticated associations, in client/server, symmetric (peer),
	and several broadcast modes. In addition to the basic NTP
	operational modes, symmetric mode and broadcast servers can
	support an interleaved mode of operation. In ntp-4.2.8p4 a bug
	was inadvertently introduced into the protocol engine that
	allows a non-authenticated zero-origin (reset) packet to reset
	an authenticated interleaved peer association. If an attacker
	can send a packet with a zero-origin timestamp and the source
	IP address of the "other side" of an interleaved association,
	the 'victim' ntpd will reset its association. The attacker must
	continue sending these packets in order to maintain the
	disruption of the association. In ntp-4.0.0 thru ntp-4.2.8p6,
	interleave mode could be entered dynamically. As of ntp-4.2.8p7,
	interleaved mode must be explicitly configured/enabled.
   Mitigation:
	Implement BCP-38.
	Upgrade to 4.2.8p11, or later, from the NTP Project Download Page
	    or the NTP Public Services Project Download Page.
	If you are unable to upgrade to 4.2.8p11 or later and have
	    'peer HOST xleave' lines in your ntp.conf file, remove the
	    'xleave' option.
	Have enough sources of time.
	Properly monitor your ntpd instances.
	If ntpd stops running, auto-restart it without -g .

1.0.2/xo/xo.c
/*
 * Copyright (c) 2014-2018, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, July 2014
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include "xo_config.h"
#include "xo.h"
#include "xo_explicit.h"

#include <getopt.h>		/* Include after xo.h for testing */

#ifndef UNUSED
#define UNUSED __attribute__ ((__unused__))
#endif /* UNUSED */

static int opt_warn;		/* Enable warnings */

static char **save_argv;
static char **checkpoint_argv;

static char *
next_arg (void)
{
    char *cp = *save_argv;

    if (cp == NULL)
	xo_errx(1, "missing argument");

    save_argv += 1;
    return cp;
}

static void
prep_arg (char *fmt)
{
    char *cp, *fp;

    for (cp = fp = fmt; *cp; cp++, fp++) {
	if (*cp != '\\') {
	    if (cp != fp)
		*fp = *cp;
	    continue;
	}

	switch (*++cp) {
	case 'n':
	    *fp = '\n';
	    break;

	case 'r':
	    *fp = '\r';
	    break;

	case 'b':
	    *fp = '\b';
	    break;

	case 'e':
	    *fp = '\e';
	    break;

	default:
	    *fp = *cp;
	}
    }

    *fp = '\0';
}

static void
checkpoint (xo_handle_t *xop UNUSED, va_list vap UNUSED, int restore)
{
    if (restore)
	save_argv = checkpoint_argv;
    else
	checkpoint_argv = save_argv;
}

/*
 * Our custom formatter is responsible for combining format string pieces
 * with our command line arguments to build strings.  This involves faking
 * some printf-style logic.
 */
static xo_ssize_t
formatter (xo_handle_t *xop, char *buf, xo_ssize_t bufsiz,
	   const char *fmt, va_list vap UNUSED)
{
    int lflag UNUSED = 0;	/* Parse long flag, though currently ignored */
    int hflag = 0, jflag = 0, tflag = 0,
	zflag = 0, qflag = 0, star1 = 0, star2 = 0;
    int rc = 0;
    int w1 = 0, w2 = 0;
    const char *cp;

    for (cp = fmt + 1; *cp; cp++) {
	if (*cp == 'l')
	    lflag += 1;
	else if (*cp == 'h')
	    hflag += 1;
	else if (*cp == 'j')
	    jflag += 1;
	else if (*cp == 't')
	    tflag += 1;
	else if (*cp == 'z')
	    zflag += 1;
	else if (*cp == 'q')
	    qflag += 1;
	else if (*cp == '*') {
	    if (star1 == 0)
		star1 = 1;
	    else
		star2 = 1;
	} else if (strchr("diouxXDOUeEfFgGaAcCsSp", *cp) != NULL)
	    break;
	else if (*cp == 'n' || *cp == 'v') {
	    if (opt_warn)
		xo_error_h(xop, "unsupported format: '%s'", fmt);
	    return -1;
	}
    }

    char fc = *cp;

    /* Handle "%*.*s" */
    if (star1)
	w1 = strtol(next_arg(), NULL, 0);
    if (star2 > 1)
	w2 = strtol(next_arg(), NULL, 0);

    if (fc == 'D' || fc == 'O' || fc == 'U')
	lflag = 1;

    if (strchr("diD", fc) != NULL) {
	long long value = strtoll(next_arg(), NULL, 0);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (strchr("ouxXOUp", fc) != NULL) {
	unsigned long long value = strtoull(next_arg(), NULL, 0);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (strchr("eEfFgGaA", fc) != NULL) {
	double value = strtold(next_arg(), NULL);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (fc == 'C' || fc == 'c' || fc == 'S' || fc == 's') {
	char *value = next_arg();
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);
    }

    return rc;
}

static void
print_version (void)
{
    fprintf(stderr, "libxo version %s%s\n",
	    xo_version, xo_version_extra);
    fprintf(stderr, "xo version %s%s\n",
	    LIBXO_VERSION, LIBXO_VERSION_EXTRA);
}

static void
print_help (void)
{
    fprintf(stderr,
"Usage: xo [options] format [fields]\n"
"    --close <path>        Close tags for the given path\n"
"    --close-instance <name> Close an open instance name\n"
"    --close-list <name>   Close an open list name\n"
"    --continuation OR -C  Output belongs on same line as previous output\n"
"    --depth <num>         Set the depth for pretty printing\n"
"    --help                Display this help text\n"
"    --html OR -H          Generate HTML output\n"
"    --json OR -J          Generate JSON output\n"
"    --leading-xpath <path> OR -l <path> "
	    "Add a prefix to generated XPaths (HTML)\n"
"    --not-first           Indicate this object is not the first (JSON)\n"
"    --open <path>         Open tags for the given path\n"
"    --open-instance <name> Open an instance given by name\n"
"    --open-list <name>   Open a list given by name\n"
"    --option <opts> -or -O <opts>  Give formatting options\n"
"    --pretty OR -p        Make 'pretty' output (add indent, newlines)\n"
"    --style <style> OR -s <style>  "
	    "Generate given style (xml, json, text, html)\n"
"    --text OR -T          Generate text output (the default style)\n"
"    --top-wrap            Generate a top-level object wrapper (JSON)\n"
"    --version             Display version information\n"
"    --warn OR -W          Display warnings in text on stderr\n"
"    --warn-xml            Display warnings in xml on stdout\n"
"    --wrap <path>         Wrap output in a set of containers\n"
"    --xml OR -X           Generate XML output\n"
"    --xpath               Add XPath data to HTML output\n");
}

static struct opts {
    int o_close_instance;
    int o_close_list;
    int o_depth;
    int o_help;
    int o_not_first;
    int o_open_instance;
    int o_open_list;
    int o_top_wrap;
    int o_version;
    int o_warn_xml;
    int o_wrap;
    int o_xpath;
} opts;

static struct option long_opts[] = {
    { "close", required_argument, NULL, 'c' },
    { "close-instance", required_argument, &opts.o_close_instance, 1 },
    { "close-list", required_argument, &opts.o_close_list, 1 },
    { "continuation", no_argument, NULL, 'C' },
    { "depth", required_argument, &opts.o_depth, 1 },
    { "help", no_argument, &opts.o_help, 1 },
    { "html", no_argument, NULL, 'H' },
    { "json", no_argument, NULL, 'J' },
    { "leading-xpath", required_argument, NULL, 'l' },
    { "not-first", no_argument, &opts.o_not_first, 1 },
    { "open", required_argument, NULL, 'o' },
    { "open-instance", required_argument, &opts.o_open_instance, 1 },
    { "open-list", required_argument, &opts.o_open_list, 1 },
    { "option", required_argument, NULL, 'O' },
    { "pretty", no_argument, NULL, 'p' },
    { "style", required_argument, NULL, 's' },
    { "text", no_argument, NULL, 'T' },
    { "top-wrap", no_argument, &opts.o_top_wrap, 1 },
    { "xml", no_argument, NULL, 'X' },
    { "xpath", no_argument, &opts.o_xpath, 1 },
    { "version", no_argument, &opts.o_version, 1 },
    { "warn", no_argument, NULL, 'W' },
    { "warn-xml", no_argument, &opts.o_warn_xml, 1 },
    { "wrap", required_argument, &opts.o_wrap, 1 },
    { NULL, 0, NULL, 0 }
};

int
main (int argc UNUSED, char **argv)
{
    char *fmt = NULL, *cp, *np;
    char *opt_opener = NULL, *opt_closer = NULL, *opt_wrapper = NULL;
    char *opt_options = NULL;
    char *opt_name = NULL;
    xo_state_t new_state = 0;
    int opt_depth = 0;
    int opt_not_first = 0;
    int opt_top_wrap = 0;
    int rc;

    argc = xo_parse_args(argc, argv);
    if (argc < 0)
	return 1;

    while ((rc = getopt_long(argc, argv, "Cc:HJl:O:o:ps:TXW",
				long_opts, NULL)) != -1) {
	switch (rc) {
	case 'C':
	    xo_set_flags(NULL, XOF_CONTINUATION);
	    break;

	case 'c':
	    opt_closer = optarg;
	    xo_set_flags(NULL, XOF_IGNORE_CLOSE);
	    break;

	case 'H':
	    xo_set_style(NULL, XO_STYLE_HTML);
	    break;

	case 'J':
	    xo_set_style(NULL, XO_STYLE_JSON);
	    break;

	case 'l':
	    xo_set_leading_xpath(NULL, optarg);
	    break;

	case 'O':
	    opt_options = optarg;
	    break;

	case 'o':
	    opt_opener = optarg;
	    break;

	case 'p':
	    xo_set_flags(NULL, XOF_PRETTY);
	    break;

	case 's':
	    if (xo_set_style_name(NULL, optarg) < 0)
		xo_errx(1, "unknown style: %s", optarg);
	    break;

	case 'T':
	    xo_set_style(NULL, XO_STYLE_TEXT);
	    break;

	case 'X':
	    xo_set_style(NULL, XO_STYLE_XML);
	    break;

	case 'W':
	    opt_warn = 1;
	    xo_set_flags(NULL, XOF_WARN);
	    break;

	case ':':
	    xo_errx(1, "missing argument");
	    break;

	case 0:
	    if (opts.o_depth) {
		opt_depth = atoi(optarg);
		
	    } else if (opts.o_help) {
		print_help();
		return 1;

	    } else if (opts.o_not_first) {
		opt_not_first = 1;

	    } else if (opts.o_xpath) {
		xo_set_flags(NULL, XOF_XPATH);

	    } else if (opts.o_version) {
		print_version();
		return 0;

	    } else if (opts.o_warn_xml) {
		opt_warn = 1;
		xo_set_flags(NULL, XOF_WARN | XOF_WARN_XML);

	    } else if (opts.o_wrap) {
		opt_wrapper = optarg;

	    } else if (opts.o_top_wrap) {
		opt_top_wrap = 1;

	    } else if (opts.o_open_list) {
		if (opt_name)
		    xo_errx(1, "only one open/close list/instance allowed: %s",
			    optarg);

		opt_name = optarg;
		new_state = XSS_OPEN_LIST;

	    } else if (opts.o_open_instance) {
		if (opt_name)
		    xo_errx(1, "only one open/close list/instance allowed: %s",
			    optarg);

		opt_name = optarg;
		new_state = XSS_OPEN_INSTANCE;

	    } else if (opts.o_close_list) {
		if (opt_name)
		    xo_errx(1, "only one open/close list/instance allowed: %s",
			    optarg);

		opt_name = optarg;
		new_state = XSS_CLOSE_LIST;

	    } else if (opts.o_close_instance) {
		if (opt_name)
		    xo_errx(1, "only one open/close list/instance allowed: %s",
			    optarg);

		opt_name = optarg;
		new_state = XSS_CLOSE_INSTANCE;

	    } else {
		print_help();
		return 1;
	    }

	    bzero(&opts, sizeof(opts)); /* Reset all the options */
	    break;

	default:
	    print_help();
	    return 1;
	}
    }

    argc -= optind;
    argv += optind;

    if (opt_options) {
	rc = xo_set_options(NULL, opt_options);
	if (rc < 0)
	    xo_errx(1, "invalid options: %s", opt_options);
    }

    xo_set_formatter(NULL, formatter, checkpoint);
    xo_set_flags(NULL, XOF_NO_VA_ARG | XOF_NO_TOP | XOF_NO_CLOSE);

    /*
     * If we have some explicit state change, handle it
     */
    if (new_state) {
	if (opt_depth > 0)
	    xo_set_depth(NULL, opt_depth);

	if (opt_not_first)
	    xo_set_flags(NULL, XOF_NOT_FIRST);

	xo_explicit_transition(NULL, new_state, opt_name, 0);
	xo_finish();
	exit(0);
    }

    fmt = *argv++;
    if (opt_opener == NULL && opt_closer == NULL && fmt == NULL) {
	print_help();
	return 1;
    }

    if (opt_top_wrap) {
	/* If we have a closing path, we'll be one extra level deeper */
	if (opt_closer && xo_get_style(NULL) == XO_STYLE_JSON)
	    opt_depth += 1;
	else
	    xo_clear_flags(NULL, XOF_NO_TOP);
    }

    if (opt_closer) {
	opt_depth += 1;
	for (cp = opt_closer; cp && *cp; cp = np) {
	    np = strchr(cp, '/');
	    if (np == NULL)
		break;
	    np += 1;
	    opt_depth += 1;
	}
    }

    if (opt_depth > 0)
	xo_set_depth(NULL, opt_depth);

    if (opt_not_first)
	xo_set_flags(NULL, XOF_NOT_FIRST);

    /* If there's an opening hierarchy, open each element as a container */
    if (opt_opener) {
	for (cp = opt_opener; cp && *cp; cp = np) {
	    np = strchr(cp, '/');
	    if (np)
		*np = '\0';
	    xo_open_container(cp);
	    if (np)
		np += 1;
	}
    }

    /* If there's an wrapper hierarchy, open each element as a container */
    if (opt_wrapper) {
	for (cp = opt_wrapper; cp && *cp; cp = np) {
	    np = strchr(cp, '/');
	    if (np)
		*np = '\0';
	    xo_open_container(cp);
	    if (np)
		*np++ = '/';	/* Put it back */
	}
    }

    /* If there's a format string, call xo_emit to emit the contents */
    if (fmt && *fmt) {
	save_argv = argv;
	prep_arg(fmt);
	xo_emit(fmt);		/* This call does the real formatting */
    }
    
    /* If there's an wrapper hierarchy, close each element's container */
    while (opt_wrapper) {
	np = strrchr(opt_wrapper, '/');
	xo_close_container(np ? np + 1 : opt_wrapper);
	if (np)
	    *np = '\0';
	else
	    opt_wrapper = NULL;
    }

    /* Remember to undo the depth before calling xo_finish() */
    opt_depth = (opt_closer && opt_top_wrap) ? -1 : 0;

    /* If there's an closing hierarchy, close each element's container */
    while (opt_closer) {
	np = strrchr(opt_closer, '/');
	xo_close_container(np ? np + 1 : opt_closer);
	if (np)
	    *np = '\0';
	else
	    opt_closer = NULL;
    }

    /* If there's a closer and a wrapper, we need to clean it up */
    if (opt_depth) {
	xo_set_depth(NULL, opt_depth);
	xo_clear_flags(NULL, XOF_NO_TOP);
    }

    /* If we're wrapping the entire content, skip the closer */
    if (opt_top_wrap && opt_opener)
	xo_set_flags(NULL, XOF_NO_TOP);

    xo_finish();

    return 0;
}
